### PR TITLE
stagger the ac and proposal height

### PR DESCRIPTION
**Pull request by lisuxiaoqi**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: libs/tendermint/consensus/consensus_commit.go [hunk 1]
```diff
@@ -255,6 +297,12 @@ func (cs *State) finalizeCommit(height int64) {
 		return
 	}
 
+	//reset offset after commitGap
+	if iavl.EnableAsyncCommit && height%commitGap == iavl.GetCommitGapOffset() {
+		log.Println("Reset commit Offset", height)
```
Comment (chengzhinei) [inline libs/tendermint/consensus/consensus_commit.go]: use cs.Logger

File: app/config/config.go [hunk 2]
```diff
@@ -1016,6 +1027,15 @@ func (c *OecConfig) SetGcInterval(value int) {
 	c.gcInterval = value
 
 }
+
+func (c *OecConfig) GetProposalACGap() int64 {
+	return c.proposalACGap
+}
+
+func (c *OecConfig) SetProposalACGap(value int64) {
```
Comment (xiangjianmeng) [inline app/config/config.go]: change SetProposalACGap to SetCommitGapOffset and positive check for value

File: libs/iavl/mutable_tree.go [hunk 3]
```diff
@@ -32,9 +32,14 @@ func SetProduceDelta(pd bool) {
...
+func SetCommitGapOffset(offset int64) {
+	CommitGapOffset = offset
+}
+
 var (
-	ignoreVersionCheck = false
-	produceDelta       = false
+	ignoreVersionCheck       = false
+	produceDelta             = false
+	CommitGapOffset    int64 = 0
```
Comment (xiangjianmeng) [inline libs/iavl/mutable_tree.go]: change CommitGapOffset to commitGapOffset because encapsulate SetCommitGapOffset GetCommitGapOffset

Comment (chengzhinei) [inline libs/iavl/mutable_tree.go]: CommitGapOffset => commitGapOffset

File: libs/tendermint/consensus/consensus_commit.go [hunk 4]
```diff
@@ -255,6 +297,12 @@ func (cs *State) finalizeCommit(height int64) {
 		return
 	}
 
+	//reset offset after commitGap
```
Comment (chengzhinei) [inline libs/tendermint/consensus/consensus_commit.go:266]: put reset logic into iavl?(when saveVersion finished)

File: libs/tendermint/consensus/consensus_commit.go [hunk 5]
```diff
@@ -242,6 +247,43 @@ func (cs *State) finalizeCommit(height int64) {
...
+			shouldOffsetAC := false
+			var i int64
+			for ; i < offset; i++ {
+				futureBPAddress := futureValidators.GetProposer().Address
+
+				// self is the validator at the offset height
+				// && nextAC happens within the offset
+				if bytes.Equal(futureBPAddress, selfAddress) {
+					// trigger ac ahead of the offset
+					shouldOffsetAC = true
```
Comment (chengzhinei) [inline libs/tendermint/consensus/consensus_commit.go]: no need use shouldOffsetAC
put "iavl.SetCommitGapOffset(i + 1)" here

File: libs/iavl/mutable_tree_oec.go [hunk 6]
```diff
@@ -76,6 +76,10 @@ func UpdateCommitGapHeight(gap int64) {
 	CommitGapHeight = gap
 }
 
+func GetCommitGapOffset() int64 {
```
Comment (chengzhinei) [inline libs/iavl/mutable_tree_oec.go]: put GetCommitGapOffset() and SetCommitGapOffset() together

File: libs/tendermint/consensus/consensus_commit.go (hunk 1 above)
Comment (xiangjianmeng) [inline libs/tendermint/consensus/consensus_commit.go]: change to logger.Info/Debug

File: libs/tendermint/consensus/consensus_commit.go [hunk 7]
```diff
@@ -242,6 +247,43 @@ func (cs *State) finalizeCommit(height int64) {
 		cs.blockExec.FireBlockTimeEvents(height, blockTime.UnixMilli(), validators.Proposer.Address)
 	}
 
+	offset := cfg.DynamicConfig.GetProposalACGap()
```
Comment (xiangjianmeng) [inline libs/tendermint/consensus/consensus_commit.go]: encapsulate a new method to do these

File: libs/tendermint/consensus/consensus_commit.go (hunk 4 above)
Comment (lisuxiaoqi) [reply libs/tendermint/consensus/consensus_commit.go:266]: it's hard to put it into iavl due to multiple iavl trees shares the same variable during the commitment

**Review by xiangjianmeng — APPROVED** [review]

**Pull request merged**